nexus_workcell_orchestrator/src/job_manager.cpp

```diff
@@ -0,0 +1,257 @@
+      auto result =
+        std::make_shared<endpoints::WorkcellRequestAction::ActionType::Result>();
+      result->success = false;
+      result->message = "Task failed";
+      job.goal_handle->abort(result);
+      break;
+    }
+    case BT::NodeStatus::SUCCESS: {
+      // Publish feedback.
+      job.task_state.status = TaskState::STATUS_FINISHED;
+      auto fb = std::make_shared<WorkcellRequest::Feedback>();
+      fb->state = job.task_state;
+      job.goal_handle->publish_feedback(std::move(fb));
+
+      auto result =
+        std::make_shared<endpoints::WorkcellRequestAction::ActionType::Result>();
+      result->success = true;
+      result->result = YAML::Dump(job.ctx->task.previous_results);
+      auto report = job.bt_logging->generate_report();
+      result->message = common::ReportConverter::to_string(report);
+      RCLCPP_INFO(job.ctx->node->get_logger(), "%s", result->message.c_str());
+      job.goal_handle->succeed(result);
+      break;
+    }
+    default: {
+      std::ostringstream oss;
+      oss << "Error ticking task [" << job.ctx->task.id <<
+        "]: Behavior tree returned invalid or unknown status [" <<
+        bt_status << "]";
+      RCLCPP_ERROR_STREAM(job.ctx->node->get_logger(), oss.str());
+      throw std::runtime_error(oss.str());
+    }
+  }
+}
+
+uint8_t JobManager::_tick_job(Job& job)
+{
+  auto& task_status = job.task_state.status;
+  switch (task_status)
+  {
+    case TaskState::STATUS_ASSIGNED: {
+      RCLCPP_INFO_THROTTLE(
+        this->_node->get_logger(),
+        *this->_node->get_clock(),
+        std::chrono::milliseconds(1000).count(),
+        "Task [%s] is pending, call \"%s\" to start this task",
+        job.task_state.task_id.c_str(),
+        endpoints::WorkcellRequestAction::action_name(
+          this->_node->get_name()).c_str());
+      break;
+    }
+    case TaskState::STATUS_QUEUED:
+    case TaskState::STATUS_RUNNING: {
+      this->_tick_bt(job);
+      break;
+    }
+    case TaskState::STATUS_FAILED: {
+      RCLCPP_WARN(
+        job.ctx->node->get_logger(), "Cannot tick task [%s] that has already failed",
+        job.ctx->task.id.c_str());
+      break;
+    }
+    case TaskState::STATUS_FINISHED: {
+      RCLCPP_WARN(
+        job.ctx->node->get_logger(), "Cannot tick task [%s] that has already finished",
+        job.ctx->task.id.c_str());
+      break;
+    }
+    default: {
+      RCLCPP_ERROR(
+        job.ctx->node->get_logger(), "Error ticking task [%s]: Unknown status [%i]",
+        job.ctx->task.id.c_str(),
+        task_status);
+      break;
+    }
+  }
+  return task_status;
+}
+
+Job& JobManager::assign_task(
+  const std::string& task_id)
+{
+  const auto it =
+    std::find_if(this->_jobs.begin(), this->_jobs.end(), [&task_id](
+        const Job& j)
+      {
+        return j.task_state.task_id == task_id;
+      });
+  if (it != this->_jobs.end())
+  {
+    throw JobError("Another task with the same id already exist");
```

Comment:
I don't mind banning exceptions, but exceptions are used in many places, I think we should do a large scale refactor if that is the direction. Also can you point me to the style guide explaining the use of exceptions? Also would be good to have a link to the style guide in a `DEVELOPERS` file.

Instead of `std::optional`, I propose to use `std::variant` instead. `std::optional` does not allow the reason for failure to be reported. `std::variant` would be closest to the way rust does error handling, an alternative is to use `std::pair`, which would be closer to the way go.

Reply:
`std::variant` works too